File: lintcode/python/0031_partition_array.py
#!/usr/bin/python -t

# two pointers

class Solution:
    """
    @param nums: The integer array you should partition
    @param k: An integer
    @return: The index after partition
    """
    def partitionArray(self, nums, k):
        # write your code here
        n = len(nums)
        l = 0
        r = n-1
        
        while l <= r:
            while l <= r and nums[l] < k:
                l += 1
            while l <= r and nums[r] >= k:
                r -= 1
            
            if l <= r:
                nums[l], nums[r] = nums[r], nums[l]
                l += 1
                r -= 1
            
        return l
        

class Solution:
    """
    @param nums: The integer array you should partition
    @param k: An integer
    @return: The index after partition
    """
    def partitionArray(self, nums, k):
        # write your code here
        n = len(nums)
        if n == 0:
            return 0
            
        index = 0
        
        for i in range(n):
            if nums[i] < k:
                nums[index], nums[i] = nums[i], nums[index]
                index += 1
                
        return index
        

File: lintcode/golang/0031_partition_array.go
/**
 * @param nums: The integer array you should partition
 * @param k: An integer
 * @return: The index after partition
 */
func partitionArray (nums []int, k int) int {
    // write your code here
    n := len(nums)
    if n == 0  || n == 1{
        return 0
    }
    
    i := 0
    j := n-1
    
    for i <= j {
        for i <= j && nums[i] < k {
            i++
        }
        
        for i <= j && nums[j] >= k {
            j--
        }
        
        if i <= j {
            nums[i], nums[j] = nums[j], nums[i]
            i++
            j--
        }
    }
    
    return i
}

